### ISSUE TITLE
實現TracIn方法來評估訓練數據影響力

### ISSUE THREAD

**Issue by sk413025** [sk413025/audio-angle-classification#42]
# 實現TracIn方法來評估訓練數據影響力

## 功能需求

實現基於論文《Estimating Training Data Influence by Tracing Gradient Descent》(Pang Wei Koh and Percy Liang)提出的TracIn方法，用於評估訓練數據對模型預測與泛化能力的影響。

### 階段1：基礎實現
- [x] 實現基本的TracIn模組，支持檢查點加載和梯度計算
- [x] 實現自影響力(self-influence)計算功能
- [x] 確保與現有訓練流程兼容

### 階段2：擴展功能
- [x] 增強對排序任務的支持(RankingTracInCP)
- [x] 增加對不同損失函數的支持(標準MRL和GHM損失)
- [x] 實現對MacOS MPS設備的支持

### 階段3：泛化能力評估
- [x] 實現訓練樣本對測試樣本影響力的計算
- [x] 優化命令行界面，支持靈活的參數配置
- [x] 整合元數據保存機制

### 階段4：分析與文檔
- [x] 生成詳細的分析報告，解釋計算原理與結果解讀
- [x] 完善README和使用文檔
- [x] 提供案例分析與應用建議

### 預期效益
1. 識別對模型表現有重大影響的訓練樣本
2. 發現模型難以處理的樣本，指導數據增強
3. 提供對模型泛化能力影響的量化分析
4. 指導訓練數據的篩選和加權

### 潛在挑戰
1. 計算資源需求較高，特別是大數據集
2. 需要儲存多個檢查點以獲得準確評估
3. 不同設備(CUDA/MPS/CPU)的兼容性問題

### 驗收標準
1. 能夠成功計算並保存自影響力與影響力分數
2. 支持多種設備和損失函數配置
3. 生成詳盡的分析報告和使用文檔
4. 與現有訓練流程無縫集成

## 參考資料
1. [原始論文](https://arxiv.org/abs/2002.08484)
2. [TracIn GitHub參考實現](https://github.com/frederick0329/TracIn)

**Comment by sk413025** [comment]
# TracIn模組實現總結報告

我們已經成功實現了TracIn訓練數據影響力評估模組，並透過PR #43將其整合到主代碼庫中。以下是實現的詳細說明：

## 關鍵功能

### 1. 核心模組實現
- **基礎TracIn類**(tracin.py)：實現論文中的基本算法，支持檢查點加載、梯度計算和影響力分數計算
- **排序特定實現**(ranking_tracin.py)：專為排序任務設計的RankingTracInCP類，支持差異梯度計算
- **命令行工具**(compute_tracin_influence.py)：靈活的命令行界面，支持各種參數配置

### 2. 多設備支持
- **CUDA GPU支持**：自動檢測並使用CUDA設備加速計算
- **MPS GPU支持**：支持MacOS的MPS硬件加速
- **CPU後備**：在無GPU環境下自動降級到CPU計算

### 3. 損失函數支持
- **標準MarginRankingLoss**：傳統排序損失函數支持
- **GHM損失**：支持與梯度調和機制(GHM)結合的損失函數
- **自定義損失**：擴展性設計支持添加新的損失函數

### 4. 影響力評估功能
- **自影響力計算**：識別訓練集中的困難樣本和異常值
- **交叉影響力**：評估訓練樣本對測試樣本預測的影響
- **批處理支持**：高效處理大量數據的批量計算

## 技術細節

### 梯度計算優化
- 實現了高效的梯度計算和存儲機制
- 支持批處理以減少內存消耗
- 針對不同設備優化了張量操作

### 元數據管理
- 保存影響力分數到JSON文件
- 支持增量更新現有元數據
- 結構化存儲便於後續分析

### 記憶體優化
- 優化梯度計算過程中的內存使用
- 及時釋放不再需要的張量
- 支持大型模型和數據集

## 使用示例

```bash
# 計算自影響力分數
python compute_tracin_influence.py \
    --frequency 500hz \
    --material plastic \
    --checkpoint-dir saved_models/model_checkpoints/your_checkpoint_dir/ \
    --compute-self-influence \
    --loss-type standard \
    --device mps

# 計算影響力分數
python compute_tracin_influence.py \
    --frequency 500hz \
    --material plastic \
    --checkpoint-dir saved_models/model_checkpoints/your_checkpoint_dir/ \
    --compute-influence \
    --num-test-samples 5 \
    --device mps
```

## 後續開發建議

1. **可視化工具**：開發專用的可視化工具，更直觀地展示影響力分數分布
2. **影響力閾值自動推薦**：基於統計方法自動推薦異常值閾值
3. **整合模型訓練**：實現在線影響力計算，與模型訓練過程整合
4. **擴展評估指標**：添加更多評估指標，如模型特徵空間中的距離等

## 參考資料

1. [原始論文: Estimating Training Data Influence by Tracing Gradient Descent](https://arxiv.org/abs/2002.08484)
2. [TracIn官方實現](https://github.com/frederick0329/TracIn)
3. [排序學習影響力分析相關研究](https://dl.acm.org/doi/10.1145/3397271.3401075)

---

此模組將幫助研究人員更好地了解模型訓練過程中樣本的重要性，為數據清理、增強和模型改進提供有價值的指導。
